pkg/test/utils/kubernetes.go

```diff
@@ -536,3 +538,37 @@ func WaitForCRDs(dClient discovery.DiscoveryInterface, crds []runtime.Object) er
 		return true, nil
 	})
 }
+
+// TestEnvironment is a struct containing the envtest environment, Kubernetes config and clients.
+type TestEnvironment struct {
+	Environment     *envtest.Environment
+	Config          *rest.Config
+	Client          client.Client
+	DiscoveryClient discovery.DiscoveryInterface
+}
+
+// StartTestEnvironment is a wrapper for controller-runtime's envtest that creates a Kubernetes API server and etcd
+// suitable for use in tests.
+func StartTestEnvironment() (env TestEnvironment, err error) {
+	env.Environment = &envtest.Environment{}
+
+	// Retry up to three times for the test environment to start up in case there is a port collision (#510).
+	for i := 0; i < 3; i++ {
```

Comment:
should this be configurable?

Reply:
Maybe, envtest already retries 5 times under the hood, so this just covers a very rare case.